net.go

```diff
@@ -155,6 +171,37 @@ func (d *Dialer) DialContext(ctx context.Context, network, address string) (net.
 	}.Build(conn), nil
 }
 
+func (d *Dialer) dialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	var lastErr error
+
+	dial := d.nd.DialContext
+	if d.testingDialContext != nil {
+		dial = d.testingDialContext
+	}
+
+	attempts := d.rt.Attempts
+	if attempts <= 0 {
+		attempts = 1
+	}
+	for i := 0; i < attempts; i++ {
+		if i > 0 {
+			time.Sleep(d.rt.Backoff)
+		}
+
+		conn, err := dial(ctx, network, address)
+		if err == nil {
+			return conn, nil
+		}
+		lastErr = err
+
+		if conn != nil {
+			conn.Close()
+		}
```

Comment:
if err != nil, conn is nil. That's the rule we were following before this patch.

Reply:
Maybe it can stay. Better safe than sorry.

Reply:
That's the rule overall but there is nothing that governs that rule.